internal/pkg/agent/application/actions/handlers/handler_helpers.go

```diff
@@ -0,0 +1,166 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package handlers
+
+import (
+	"context"
+	"time"
+
+	"golang.org/x/sync/errgroup"
+
+	"github.com/elastic/elastic-agent-client/v7/pkg/client"
+	"github.com/elastic/elastic-agent-libs/logp"
+
+	"github.com/elastic/elastic-agent/internal/pkg/agent/application/coordinator"
+	"github.com/elastic/elastic-agent/internal/pkg/agent/errors"
+	"github.com/elastic/elastic-agent/internal/pkg/core/backoff"
+	"github.com/elastic/elastic-agent/internal/pkg/fleetapi"
+	"github.com/elastic/elastic-agent/pkg/component"
+)
+
+type actionCoordinator interface {
+	State() coordinator.State
+	PerformAction(ctx context.Context, comp component.Component, unit component.Unit, name string, params map[string]interface{}) (map[string]interface{}, error)
+}
+
+type upgradeCoordinator interface {
+	actionCoordinator
+	Upgrade(ctx context.Context, version string, sourceURI string, action *fleetapi.ActionUpgrade, skipVerifyOverride bool, pgpBytes ...string) error
+}
+
+type performActionFunc func(context.Context, component.Component, component.Unit, string, map[string]interface{}) (map[string]interface{}, error)
+
+type dispatchableAction interface {
+	MarshalMap() (map[string]interface{}, error)
+	Type() string
+}
+
+type unitWithComponent struct {
+	unit      component.Unit
+	component component.Component
+}
+
+func findMatchingUnitsByActionType(state coordinator.State, typ string) []unitWithComponent {
+	ucs := make([]unitWithComponent, 0)
+	for _, comp := range state.Components {
+		if comp.Component.InputSpec != nil && contains(comp.Component.InputSpec.Spec.ProxiedActions, typ) {
+			name := comp.Component.InputSpec.Spec.Name
```

Comment:
As of [today's checkin](https://github.com/elastic/elastic-agent/pull/2952) you can get this value directly from the component with `comp.Component.InputType` instead of `comp.Component.InputSpec.Spec.Name`.

Checks of `unit.Config.Type` below are redundant since all inputs in a component must have the same type -- you could remove those checks and add a comment to this effect, or keep it as is and comment with `// All input units should match the component input type, but let's be cautious` or something ^^

Reply:
Updated.